app/src/marketplace/containers/EditProductPage2/EditControllerProvider.jsx

```diff
@@ -67,19 +70,29 @@ function useEditController(product: Product) {
     ])
 
     const save = useCallback(async () => {
+        if (!originalProduct) { throw new Error('originalProduct is missing') }
         const savedSuccessfully = await savePending.wrap(async () => {
-            // save product
-            await putProduct(product, product.id || '')
+            const nextProduct = {
+                ...product,
+            }
 
             // upload image
             if (product.newImageToUpload != null) {
                 try {
-                    await postImage(product.id || '', product.newImageToUpload)
+                    const { imageUrl: newImageUrl, thumbnailUrl: newThumbnailUrl } = await postImage(product.id || '', product.newImageToUpload)
+                    nextProduct.imageUrl = newImageUrl
+                    nextProduct.thumbnailUrl = newThumbnailUrl
+                    delete nextProduct.newImageToUpload
                 } catch (e) {
                     console.error('Could not upload image', e)
                 }
             }
 
+            // save product
+            await putProduct(State.update(originalProduct, () => ({
```

Comment:
Should this abort if it unmounted between `await postImage(product.id || '', product.newImageToUpload)` and here? Maybe not, but perhaps put in an explicit comment about not aborting if so.